Sources/Redis/Application+Redis.swift

```diff
@@ -0,0 +1,72 @@
+import Vapor
+
+extension Application {
+    public var redis: Redis {
+        .init(application: self)
+    }
+
+    public struct Redis {
+        struct ConfigurationKey: StorageKey {
+            typealias Value = RedisConfiguration
+        }
+
+        public var configuration: RedisConfiguration? {
+            get {
+                self.application.storage[ConfigurationKey.self]
+            }
+            nonmutating set {
+                self.application.storage[ConfigurationKey.self] = newValue
+            }
+        }
+
+
+        struct PoolKey: StorageKey, LockKey {
+            typealias Value = EventLoopGroupConnectionPool<RedisConnectionSource>
+        }
+
+        internal var pool: EventLoopGroupConnectionPool<RedisConnectionSource> {
+            if let existing = self.application.storage[PoolKey.self] {
+                return existing
+            } else {
+                let lock = self.application.locks.lock(for: PoolKey.self)
+                lock.lock()
+                defer { lock.unlock() }
+                guard let configuration = self.configuration else {
+                    fatalError("Redis not configured. Use app.redis.configuration = ...")
+                }
+                let new = EventLoopGroupConnectionPool(
+                    source: RedisConnectionSource(configuration: configuration, logger: self.application.logger),
+                    maxConnectionsPerEventLoop: 1,
+                    logger: self.application.logger,
+                    on: self.application.eventLoopGroup
+                )
+                self.application.storage.set(PoolKey.self, to: new) {
+                    $0.shutdown()
+                }
+                return new
+            }
+        }
+
+        let application: Application
+    }
+}
+
+extension Application.Redis: RedisClient {
+    public var logger: Logger? {
+        self.application.logger
+    }
+    
+    public var eventLoop: EventLoop {
+        self.application.eventLoopGroup.next()
+    }
+    
+    public func send(command: String, with arguments: [RESPValue]) -> EventLoopFuture<RESPValue> {
+        self.application.redis.pool.withConnection(
+            logger: logger,
+            on: nil
```

Comment:
I think it might be better to leave this to nil instead of just selecting the, random in practice, next eventLoop of the `Application` `EventLoopGroup` since, unless I'm mistaken, it will avoid a potentially useless context switch in practice.

Reply:
Yup this is fine. `EventLoopGroupConnectionPool` is going to do `self.eventLoopGroup.next()` internally anyway to get an EL. Context switching is unavoidable here really since application is used on the main thread and none of the event loops are tied to the main thread.